src/Generator/TypeGenerator.php

```diff
@@ -273,26 +274,93 @@ protected function generateInputFields(array $config): string
+     *
+     * @throws GeneratorException
+     */
     protected function generateExtraCode(array $value, string $key, ?string $argDefinitions = null, string $default = 'null', array &$compilerNames = null): string
     {
         $resolve = $value['resolve'] ?? false;
+        $groups = $value['validationGroups'] ?? null;
         $extraCode = '';
 
-        if ('resolve' === $key && $resolve && (false !== \strpos($resolve->__toString(), 'validator'))) {
-            $compilerNames[] = 'validator';
+        // Generate validation code for the resolver callback
+        if ('resolve' === $key) {
+            $autoValidation = true;
+            $autoThrow = true;
+
             $mapping = $this->buildValidationMapping($value);
-            $extraCode .= $this->generateValidation($mapping);
-            $this->addInternalUseStatement(InputValidator::class);
+
+            // If `validator` injected
+            if (false !== \strpos($resolve->__toString(), 'validator')) {
+                $compilerNames[] = 'validator';
+                $autoValidation = false;
+            }
+
+            // If `errors` injected
+            if (false !== \strpos($resolve->__toString(), 'errors')) {
+                $compilerNames[] = 'errors';
+                $autoThrow = false;
+                $this->addUseStatement(ResolveErrors::class);
+
+                $extraCode .= '$errors = new ResolveErrors();'."\n\n<spaces><spaces>";
+            }
+
+            if ($mapping) {
+                $extraCode .= $this->generateValidation($mapping, $autoValidation, $autoThrow, $groups);
+                $this->addInternalUseStatement(InputValidator::class);
+            } elseif (false === $autoValidation) {
+                throw new GeneratorException(
+                    'Unable to inject an instance of the InputValidator. No validation constraints provided. '.
+                    'Please remove the InputValidator argument from the list of dependencies of your '.
+                    'resolver or provide validation configs.'
+                );
+            }
         }
 
         return $extraCode;
     }
 
-    protected function generateValidation(array $rules): string
+    /**
+     * Generates validation code in the resolver callback.
+     *
+     * @param array      $rules
+     * @param bool       $autoValidation
+     * @param bool       $autoThrow
+     * @param array|null $groups
+     *
+     * @return string
+     */
+    protected function generateValidation(array $rules, bool $autoValidation, bool $autoThrow, ?array $groups): string
     {
         $code = $this->processTemplatePlaceHoldersReplacements('ValidatorCode', $rules, self::DEFERRED_PLACEHOLDERS);
         $code = $this->prefixCodeWithSpaces($code, 2);
 
+        if ($autoValidation) {
+            $code .= "\n\n<spaces><spaces>";
+
+            if ($groups) {
+                $groups = '['.\implode(',', \array_map(function ($value) { return "'$value'"; }, $groups)).']';
```

Comment:
why not using `var_export` or `json_encode` here?

Reply:
Didn't use `var_export` because it doesn't give a good control over the formatting.

With `var_export` it would output following:
```php
$validator->validate(array (
  0 => 'groups1',
  1 => 'my_group',
));
```
but with my code it outputs:
```php
$validator->validate(['groups1','my_group']);
```
which is more consistent with the rest of the code in the generated classes.

What does `json_encode` have to do with this case?

Reply:
Changed to `json_encode`